basic_pipeline/3_Source.md

```diff
@@ -0,0 +1,132 @@
+    action = [buffer: {:output, %Buffer{payload: chosen}}]
+    action = if size > 1, do: action ++ [redemand: :output], else: action
+    {{:ok, action}, state}
+  end
+ end
+ ...
+end
+
+```
+
+The first callback's definition matches the situation in which 0 buffers of data are requested - then we simply do nothing. You might wonder why would anybody like to request for 0 buffers - and you will find out soon that it will be us who will do so!
+The second callback's body describes the situation in which some buffers were in fact requested (and therefore the `size` argument is positive). Then we are checking if we have any packets left in the list persisting in the state of the element. If that list is empty, we are sending an `end_of_stream` action, indicating that there will be no more buffers sent through the `:output` pad and that is why there is no point in requesting more buffers.
+However, in case of the `content` list of packets being non-empty, we are taking the head of that list, and storing the remaining tail of the list in the state of the element. Later on, we are defining the actions we want to take - that is, we want to return a buffer with the head packet from the original list. We make use of the [`buffer:` action](https://hexdocs.pm/membrane_core/Membrane.Element.Action.html#t:buffer_t/0), and specify that we want to transmit the [`%Buffer`](https://hexdocs.pm/membrane_core/Membrane.Buffer.html#t:t/0) structure through the `:output` pad. Note the fields available in the `%Buffer` structure - in our case, we make use of only the `:payload` fields, which, according to the documentation, can be of `any` type - however, in almost all cases you will need to send binary data within this field. Any structured data (just like timestamps etc.) should be passed in the other fields available in the `%Buffer`, designed especially for that cases.
+However, there is the other action that is optionally taken if the requested `size` (meaning - the number of buffers requested) is greater than 1 (which basically means that the request cannot be completely fulfilled in the single `handle_demand` invocation we are just processing), we specify the `:redemand` action to take place on the `:output` pad. This action will simply queue the `handle_demand/4` callback to be invoked once again. The great thing here is that the `size` of the demand will be automatically determined by the element and we do not need to specify it anyhow. Redemanding, in the context of sources, helps use simplify the logic of the `handle_demand` callback since all we need to do in that callback is to supply a single piece of data and in case this is not enough, take a [`:redemand`](https://hexdocs.pm/membrane_core/Membrane.Element.Action.html#t:redemand_t/0) action and invoke that callback once again. As you will see later, the process of redemanding is even more powerful in the context of the filter elements, as we will take advantage of the fact that it is asynchronous. 
```

Comment:
Remember to edit this when you finally remove the redundant `if` from above.

Reply:
done